nemo/utils/app_state.py

```diff
@@ -45,6 +45,12 @@ def __init__(self):
         self._create_checkpoint_callback = None
         self._checkpoint_callback_params = None
 
+        # Save and Restore (.nemo)
+        self._tmpdir_name = None
+        self._model_config_yaml = "model_config.yaml"
```

Comment:
This logic assumes that there is a single model per application.
We currently don't have other use-cases but in future that assumption will break when we will need to add teacher-student training - in those cases it is likely to have 2 models per application.

Reply:
Good point. I think we can generalize it to multiple models when we add teacher-student.

Reply:
Yes AppState in general always (rightly) assumes only 1 model is being constructed ever in entire runtime - because its dealing with global defaults. This setup here will need to change to support more than 1 model being restored.

Reply:
So we have two options:

1. Re-design this part of this PR now.
2. Re-design this part when we start adding teacher-student

I think both options are OK.

Reply:
I'll go with 2 as this PR is already too big.

Reply:
(2) is fine with me too.